extensions/stackdriver/metric/record.cc

```diff
@@ -244,29 +244,82 @@ void addHttpSpecificTags(const ::Wasm::Common::RequestInfo& request_info,
                        std::to_string(response_code));
 }
 
+TagKeyValueList getMetricTagMap(
+    TagKeyValueList input_map,
+    std::vector<std::pair<std::string, std::string>> overrides) {
+  if (overrides.size() == 0) {
+    return input_map;
+  }
+
+  TagKeyValueList out;
+  for (auto in : input_map) {
+    auto it = std::find_if(overrides.begin(), overrides.end(),
+                           [&in](const auto& override) {
+                             return override.first == in.first.name();
+                           });
+    if (it != overrides.end()) {
+      auto tag_key = opencensus::tags::TagKey::Register(it->first);
+      out.emplace_back(tag_key, it->second);
+    } else {
+      out.emplace_back(in.first, in.second);
+    }
+  }
+
+  auto it = std::find_if(
+      overrides.begin(), overrides.end(),
+      [](const auto& override) { return override.first == "api_version"; });
+  if (it != overrides.end()) {
+    out.emplace_back(apiVersionKey(), it->second);
+  }
+
+  it = std::find_if(
+      overrides.begin(), overrides.end(),
+      [](const auto& override) { return override.first == "api_name"; });
+  if (it != overrides.end()) {
+    out.emplace_back(apiNameKey(), it->second);
+  }
+
+  return out;
+}
+
 }  // namespace
 
 void record(bool is_outbound, const ::Wasm::Common::FlatNode& local_node_info,
             const ::Wasm::Common::FlatNode& peer_node_info,
             const ::Wasm::Common::RequestInfo& request_info,
-            bool record_http_size_metrics) {
+            bool record_http_size_metrics,
+            std::unordered_map<std::string,
+                               std::vector<std::pair<std::string, std::string>>>
+                overrides) {
   double latency_ms = request_info.duration /* in nanoseconds */ / 1000000.0;
   if (is_outbound) {
     TagKeyValueList tagMap =
         getOutboundTagMap(local_node_info, peer_node_info, request_info);
     addHttpSpecificTags(request_info, tagMap);
+
     if (record_http_size_metrics) {
       opencensus::stats::Record(
-          {{clientRequestCountMeasure(), 1},
-           {clientRoundtripLatenciesMeasure(), latency_ms},
-           {clientRequestBytesMeasure(), request_info.request_size},
-           {clientResponseBytesMeasure(), request_info.response_size}},
-          tagMap);
+          {{clientRequestCountMeasure(), 1}},
```

Comment:
Record is expensive. It will make a copy of the tagmap: https://github.com/census-instrumentation/opencensus-cpp/blob/afe0460f92fb78e6d6cf8c8a30ced9bc5e2e57d3/opencensus/stats/recording.h#L44.

We should only call with separated record function if the metric has customization.

Reply:
That's gonna be tough, given the structure of the code. I'll see what can be done here.

Reply:
I'm at a loss, given that `Record()` only takes initializer lists.  At best, if there are no customizations, we could use the old code path, but pay the penalty if there are _any_ customizations. Thoughts?

Reply:
Yeah that sgtm.

Reply:
updated. please review to make sure i haven't borked anything in the rewrite.